zkevm-circuits/src/circuit_tools/cached_region.rs

```diff
@@ -0,0 +1,262 @@
+    pub fn get_fixed(&self, _row_index: usize, _column_index: usize, _rotation: Rotation) -> F {
+        unimplemented!("fixed column");
+    }
+
+    pub fn get_advice(&self, row_index: usize, column_index: usize, rotation: Rotation) -> F {
+        self.advice[column_index - self.width_start]
+            [(((row_index - self.height_start) as i32) + rotation.0) as usize]
+    }
+
+    pub fn region(&mut self) -> &mut Region<'_, F> {
+        self.region
+    }
+
+    pub fn challenges(&self) -> &Challenges<Value<F>> {
+        self.challenges
+    }
+
+    pub fn word_rlc(&self, n: U256) -> Value<F> {
+        self.challenges
+            .evm_word()
+            .map(|r| rlc::value(&n.to_le_bytes(), r))
+    }
+    pub fn empty_code_hash_rlc(&self) -> Value<F> {
+        self.word_rlc(CodeDB::empty_code_hash().to_word())
+    }
+
+    /// Constrains a cell to have a constant value.
+    ///
+    /// Returns an error if the cell is in a column where equality has not been
+    /// enabled.
+    pub fn constrain_constant<VR>(
+        &mut self,
+        cell: AssignedCell<F, F>,
+        constant: VR,
+    ) -> Result<(), Error>
+    where
+        VR: Into<Assigned<F>>,
+    {
+        self.region.constrain_constant(cell.cell(), constant.into())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct StoredExpression<F, T: CustomTable> {
+    pub(crate) name: String,
+    cell: Cell<F>,
+    cell_type: CellType_<T>,
+    expr: Expression<F>,
+    expr_id: String,
+}
+
+impl<F, T: CustomTable> Hash for StoredExpression<F, T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.expr_id.hash(state);
+        self.cell_type.hash(state);
+    }
+}
+
+impl<F: Field, T: CustomTable> StoredExpression<F, T>  {
+    pub fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+    ) -> Result<Value<F>, Error> {
+        let value = self.expr.evaluate(
+            &|scalar| Value::known(scalar),
+            &|_| unimplemented!("selector column"),
+            &|fixed_query| {
+                Value::known(region.get_fixed(
+                    offset,
+                    fixed_query.column_index(),
+                    fixed_query.rotation(),
+                ))
+            },
+            &|advice_query| {
+                Value::known(region.get_advice(
+                    offset,
+                    advice_query.column_index(),
+                    advice_query.rotation(),
+                ))
+            },
+            &|_| unimplemented!("instance column"),
+            &|challenge| *region.challenges().indexed()[challenge.index()],
+            &|a| -a,
+            &|a, b| a + b,
+            &|a, b| a * b,
+            &|a, scalar| a * Value::known(scalar),
+        );
+        self.cell.assign_value(region, offset, value)?;
+        Ok(value)
+    }
+}
+
+/// Returns the random linear combination of the inputs.
+/// Encoding is done as follows: v_0 * R^0 + v_1 * R^1 + ...
+pub(crate) mod rlc {
```

Comment:
Yea, I put it here because it's being called [here](https://github.com/CeciliaZ030/zkevm-circuits/blob/a04940ce8acc00559ca920d33f4bd10de25c5866/zkevm-circuits/src/circuit_tools/cached_region.rs#L142)